src/market.py

```diff
@@ -0,0 +1,160 @@
+"""Market module to interact with Serum DEX."""
+import base64
+from typing import Any
+
+from construct import Bytes, Int8ul, Int64ul, Padding  # type: ignore
+from construct import Struct as cStruct  # type: ignore
+from solana.publickey import PublicKey
+from solana.rpc.api import Client
+
+from .layouts.account_flags import ACCOUNT_FLAGS_LAYOUT
+
+DEFAULT_DEX_PROGRAM_ID = PublicKey(
+    "4ckmDgGdxQoPDLUkDT3vHgSAkzA3QRdNq5ywwY4sUSJn",
+)
+
+MARKET_FORMAT = cStruct(
+    Padding(5),
+    "account_flags" / ACCOUNT_FLAGS_LAYOUT,
+    "own_address" / Bytes(32),
+    "vault_signer_nonce" / Int64ul,
+    "base_mint" / Bytes(32),
+    "quote_mint" / Bytes(32),
+    "base_vault" / Bytes(32),
+    "base_deposits_total" / Int64ul,
+    "base_fees_accrued" / Int64ul,
+    "quote_vault" / Bytes(32),
+    "quote_deposits_total" / Int64ul,
+    "quote_fees_accrued" / Int64ul,
+    "quote_dust_threshold" / Int64ul,
+    "request_queue" / Bytes(32),
+    "event_queue" / Bytes(32),
+    "bids" / Bytes(32),
+    "asks" / Bytes(32),
+    "base_lot_size" / Int64ul,
+    "quote_lot_size" / Int64ul,
+    "fee_rate_bps" / Int64ul,
+    Padding(7),
+)
+
+# TODO: probably need to change the amount of padding since they recently changed it.
+# See here: https://github.com/project-serum/serum-js/commit/87c25716c0f2f1092cf27467dd8bb06aabb83fdb
+MINT_LAYOUT = cStruct(Padding(36), "decimals" / Int8ul, Padding(3))
+
+
+class Market:
+    """Represents a Serum Market."""
+
+    _decode: Any
+    _baseSplTokenDecimals: int
+    _quoteSolTokenDecimals: int
+    _skipPreflight: bool
+    _confirmations: int
+    _porgram_id: PublicKey
+
+    # pylint: disable=too-many-arguments
+    def __init__(
+        self,
+        decoded: Any,
+        base_mint_decimals: int,
+        quote_mint_decimals: int,
+        options: Any,  # pylint: disable=unused-argument
+        program_id: PublicKey = DEFAULT_DEX_PROGRAM_ID,
+    ):
+        # TODO: add options
+        if not decoded.account_flags.initialized or not decoded.account_flags.market:
+            raise Exception("Invalid market state")
+        self._decode = decoded
+        self._base_spl_token_decimals = base_mint_decimals
+        self._quote_spl_token_decimals = quote_mint_decimals
+        self._skip_preflight = False
+        self._confirmations = 10
+        self._program_id = program_id
+
+    @staticmethod
+    # pylint: disable=unused-argument
+    def load(endpoint: str, market_address: str, options: Any, program_id: PublicKey = DEFAULT_DEX_PROGRAM_ID):
```

Comment:
Let's abstract the endpoint and have the user pick from `localnet`, `testnet`, `devnet` and `mainnet`.

Reply:
Discussed over Discord, agreed that we will visit this later. This is mainly due to the fact that developer might want more flexibility, or in the future, say there might be a proxy, e.g. Infura which front the Solana.